File: LinkedList/SinglyLinkedList/Medium/Yintersection.java
package Medium;

import x.*;

public class Yintersection {
    public static void main(String[] args) {
        LinkedList l1 = new LinkedList();
        LinkedList l2 = new LinkedList();

        l1.append(1);
        l1.append(2);
        l1.append(3);
        l1.append(4);

        l2.append(5);
        l2.append(6);
        l2.head.next.next = l1.head.next.next; // 5

        System.out.println(findYIntersection(l1, l2));
    }

    static int findYIntersection(LinkedList l1, LinkedList l2) {
        int len1 = l1.length();
        int len2 = l2.length();
        int d = (len1 > len2) ? len1 - len2 : len2 - len1;

        Node cur1 = l1.head;
        Node cur2 = l2.head;

        if (len1 > len2) {
            // iterate over the bigger list l1
            for (int i = 0; i < d; i++) {
                cur1 = cur1.next;
            }
        } else {
            for (int i = 0; i < d; i++) {
                cur2 = cur2.next;
            }
        }

        // cur1 and cur2 are updated
        while (cur1 != null && cur2 != null) {
            if (cur1 == cur2) {
                return cur1.val;

            }
            cur1 = cur1.next;
            cur2 = cur2.next;
        }

        return -1;
    }
}

File: LinkedList/SinglyLinkedList/Medium/YIntersection.js
import { Node, LinkedList } from "../x/LinkedList.js";

const list1 = new LinkedList([10, 15, 30]);
const list2 = new LinkedList([3, 6, 9, 20, 1]);
list1.head.next.next.next = list2.head.next.next.next;

list1.print(); // 3, 6, 9, 15, 30
list2.print(); // 10, 15, 30

// answer should be 15
// making lists of same size

const yIntersection = (list1, list2) => {
    const len1 = list1.length;
    const len2 = list2.length;
    const diff = len1 - len2;

    let cur1 = list1.head;
    let cur2 = list2.head;

    if (len1 > len2) {
        for (let i = 0; i < diff; i++) {
            cur1 = cur1.next;
        }
    } else {
        for (let i = 0; i < diff; i++) {
            cur2 = cur2.next;
        }
    }

    // iterate both list
    while (cur1 != null && cur2 != null) {
        if (cur1 === cur2) {
            console.log("Y intersection found! at " + cur1.val);
            break;
        }
        cur1 = cur1.next;
        cur2 = cur2.next;
    }
};
yIntersection(list1, list2);
